ludwig/encoders/image_encoders.py

```diff
@@ -315,105 +244,783 @@ def output_shape(self) -> torch.Size:
+#     @property
+#     def input_shape(self) -> torch.Size:
+#         return torch.Size(self._input_shape)
+#
+#     @property
+#     def output_shape(self) -> torch.Size:
+#         return torch.Size(self._output_shape)
+
+
+class TVBaseEncoder(Encoder):
     def __init__(
         self,
-        height: int,
-        width: int,
-        num_channels: int = 3,
+        model_variant: Union[str, int] = None,
         use_pretrained: bool = True,
-        pretrained_model: str = "google/vit-base-patch16-224",
         saved_weights_in_checkpoint: bool = False,
-        hidden_size: int = 768,
-        num_hidden_layers: int = 12,
-        num_attention_heads: int = 12,
-        intermediate_size: int = 3072,
-        hidden_act: str = "gelu",
-        hidden_dropout_prob: float = 0.1,
-        attention_probs_dropout_prob: float = 0.1,
-        initializer_range: float = 0.02,
-        layer_norm_eps: float = 1e-12,
-        gradient_checkpointing: bool = False,
-        patch_size: int = 16,
+        model_cache_dir: Optional[str] = None,
         trainable: bool = True,
-        output_attentions: bool = False,
-        encoder_config=None,
         **kwargs,
     ):
-        """Creates a ViT encoder using transformers.ViTModel.
-
-        use_pretrained: If True, uses a pretrained transformer based on the
-            pretrained_model argument.
-        pretrained: If str, expects the path to a pretrained model or the id of
-            a model on huggingface.co, and ignores the configuration provided in
-            the arguments.
-        """
         super().__init__()
-        self.config = encoder_config
-
-        try:
-            from transformers import ViTConfig, ViTModel
-        except ModuleNotFoundError:
-            raise RuntimeError(
-                " transformers is not installed. "
-                "In order to install all image feature dependencies run "
-                "pip install ludwig[image]"
-            )
 
+        logger.debug(f" {self.name}")
         # map parameter input feature config names to internal names
-        img_height = height
-        img_width = width
-        in_channels = num_channels
+        self.model_variant = model_variant
+        self.use_pretrained = use_pretrained
+        self.model_cache_dir = model_cache_dir
 
-        img_width = img_width or img_height
-        if img_width != img_height:
-            raise ValueError("img_height and img_width should be identical.")
-        self._input_shape = (in_channels, img_height, img_width)
+        # remove any Ludwig specific keyword parameters
+        kwargs.pop("encoder_config", None)
+        kwargs.pop("type", None)
+
+        # cache pre-trained models if requested
+        # based on https://github.com/pytorch/vision/issues/616#issuecomment-428637564
+        if self.model_cache_dir is not None:
+            os.environ["TORCH_HOME"] = self.model_cache_dir
+
+        model_id = f"{self.torchvision_model_type}-{self.model_variant}"
```

Comment:
Instead of keying by and reconstructing this special `model_id` key in multiple places, it could be simpler to make the registry a two-level structure, i.e. `Dict[torchvision_model_type, Dict[variant_id, TVModelVariant]]`.

```python
# For list of variant IDs (str) for a given model type.
torchvision_model_registry[model_type].keys()

# For TVModelVariants for a given model type.
torchvision_model_registry[model_type].values()

# For a list of torchvision model types.
torchvision_model_registry.keys()
```

This would also remove the need for separate per-model-type global variables that you are referencing in tests.

Reply:
@justinxzhao   I like this idea.  

I believe a few of the earlier comments re: `TVModelVariant` and `TVVariantSpec` are related to this comment.  If these related comments are implemented, is this how you expect registration will occur for the pretrained models?

```
@register_torchvision_variants(
    "convnext_torch",
    {
        "tiny": TVModelVariant(...),
        "small": TVModelVariant(...),
        "base": TVModelVariant(...),
       "large": TVModelVariant(...)
   }
) 
@register_encoder("convnext_torch", IMAGE)
class ConvNextEncoder(TVBaseEncoder):
...
```

Reply:
Glad that you think it's a good idea!

Here's one way I imagine this could work. You continue to register variants as a list, and the map building happens within the registration function.

The registry's  key is derived from `cls.torchvision_model_type` while the nested dict's key comes from the `TVModelVariant.variant_id` directly. 

```python
def register_torchvision_model_variants(variants: List[TVModelVariant]):
    def wrap(cls):
        for variant in variants:
            torchvision_model_registry[cls.torchvision_model_type][variant.variant_id] = variant  # kest buy
        return cls
    return wrap


@register_torchvision_model_variants([
    TVModelVariant(variant_id="tiny", create_model_function=tvm.convnext_tiny, model_weights=tvm.convnext.ConvNeXt_Tiny_Weights),
    TVModelVariant(variant_id="small", create_model_function=tvm.convnext_small, model_weights=tvm.convnext.ConvNeXt_Small_Weights),
    TVModelVariant(variant_id="base", create_model_function=tvm.convnext_base, model_weights=tvm.ConvNeXt_Base_Weights),
    TVModelVariant(variant_id="large", create_model_function=tvm.convnext_large, model_weights=tvm.ConvNeXt_Large_Weights),
])
@register_encoder("convnext_torch", IMAGE)
class TVConvNeXtEncoder(TVBaseEncoder):
    # ...
    @property
    def get_torchvision_model_type(self):
        return "convnext_torch"
    # ...
  


# For TVModelVariants for a given model type.
torchvision_model_registry["convnext_torch"].values()
    # Would return a list of 4 TVModelVariants pertaining to "convnext_torch".

# For a list of torchvision model types.
torchvision_model_registry.keys()
    # Would return a list the names of all 4 variants relevant to "convnext_torch", i.e. ["tiny", "small", "base", "large].
```

Reply:
implemented the recommendation